### PR TITLE
Fix race condition in LSP FindAllReferences when linked files were involved.

### PR DESCRIPTION
**Pull request by dibarbet**
Fixes https://github.com/dotnet/vscode-csharp/issues/7391

TestFindAllReferencesAsync_LinkedFile should fail a few times (other passes)

### CODE REVIEW

File: src/LanguageServer/Protocol/Handler/BufferedProgress.cs [hunk 1]
```diff
@@ -43,14 +43,17 @@ public void Dispose()
         /// </summary>
         public void Report(T value)
         {
-            // Don't need to lock _underlyingProgress.  It is inherently thread-safe itself being an IProgress implementation.
-            _underlyingProgress?.Report(value);
-
-            if (_buffer != null)
+            lock (_buffer)
```
Comment (333fred) [inline src/LanguageServer/Protocol/Handler/BufferedProgress.cs:46]: This will null-ref if `_buffer` is null, which is checked below. What's the purpose?

Comment (CyrusNajmabadi) [reply src/LanguageServer/Protocol/Handler/BufferedProgress.cs:46]: this PR is not for review.  it's sharing between me and david as i isolate which part of the stack has the issue.

**Review by 333fred — APPROVED** [review]

**Comment by CyrusNajmabadi** [comment]
Introduced in https://github.com/dotnet/roslyn/pull/73613.

File: src/LanguageServer/Protocol/Handler/References/FindUsagesLSPContext.cs [hunk 2]
```diff
@@ -138,15 +138,15 @@ await foreach (var reference in references)
                     // Each reference should be associated with a definition. If this somehow isn't the
                     // case, we bail out early.
                     if (!_definitionToId.TryGetValue(reference.Definition, out var definitionId))
-                        return;
+                        continue;
```
Comment (ToddGrun) [inline src/LanguageServer/Protocol/Handler/References/FindUsagesLSPContext.cs:141]: >continue [](http://example.com/codeflow?start=24&length=8)

should the comment be updated?

Comment (CyrusNajmabadi) [reply src/LanguageServer/Protocol/Handler/References/FindUsagesLSPContext.cs:141]: enh

we're still 'bailing out' on that reference.

**Review by ToddGrun — APPROVED** [review]
:shipit:

**Pull request merged**